ships/0035-beta-api-changes.md

```diff
@@ -0,0 +1,158 @@
+4. In the `BuildStrategy` resource(_namespace and cluster scope_), replace the `.spec.buildSteps[]`items with customize fields, instead of the whole `corev1.Container` Go type.
+5. In the `Build` resource, replace the `build.build.dev/build-run-deletion` annotation in favor of `.spec.retention.atBuildDeletion`.
+6. In the `BuildRun` resource, rename `.status.latestTaskRunRef` to `.status.taskRunName`.
+7. In the `BuildRun` resource, consolidate `.spec.buildRef.name` and `.spec.buildSpec` into `.spec.build.name` and `.spec.build.spec`.
+8. Use `type discriminators` if an API object has some form of implied polymorphism. For example, source types or volume sources.
+
+### Deprecation
+
+1. In the `Build` resource, deprecate `.spec.sources`. We only allow `.spec.source`.
+2. In the `Build` resource, deprecate `.spec.dockerfile`.
+3. In the `Build` resource, deprecate `.spec.builder`.
+4. In the `Build` resource, deprecate `.spec.volumes[].description`.
+5. In the `Build`, `BuildStrategy` and `ClusterBuildStrategy` resource, remove the `status` subresource.
+6. In the `BuildRun` resource, remove the support for auto-generating a service account.
+
+### Supported Use Cases
+
+#### Story 1
+
+As a user, I can create a Shipwright _v1beta1_ CRD. Shipwright should be able to convert this into a _v1alpha1_ CRD and stored that version in ETCD.
+
+#### Story 2
+
+As an external Kubernetes controller, I should be able to operate on Shipwright _v1beta1_ or _v1alpha1_ CRD's, independently of the API stored version.
+
+### Implementation Details/Notes/Constraints [optional]
+
+We rely on the controller-runtime conversion model known as "hub and spoke". In this model, we will mark one version as the "hub", and all other versions just define conversion to and from the hub. In the scenario of needing to convert between two non-hub versions, we first convert to the hub version, and from there to the desired version:
+
+
+![Controller-runtime hub and spoke model](assets/0006-hub-and-spoke-model.png)
+
+For requests to a particular version that is not the API stored version in ETCD, we will require a webhook. The webhook  will be call out by the API server to do the conversion. For the webhook, we will rely on the [webhook](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/webhook/conversion) package from controller-runtime. The [conversion](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/conversion) package will provide us the interface definitions that an API type must implement to be supported by the generic conversion webhook from controller-runtime.
+
+From the above, the implementation can be categorize in two areas:
+
+
+#### Webhook Implementation
+
+1. We need to implement a webhook, following the [webhook](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/webhook/conversion) controller-runtime package. This webhook should assume that the related API types implement the conversion interface definitions.
```

Comment:
I think we should decide in the proposal if we make the webhook a separate deployment, or serve it in the existing deployment. I guess we go with the first option also thinking forward to eventually contain admission webhook endpoints.

Reply:
Its also my preferred option, I'm making this clear in the above paragraph.